### ISSUE TITLE
Computed cached data appears in spy reports

### ISSUE THREAD

**Issue by kubk** [mobxjs/mobx#3109]
Hello, I am writing a logging library that works with Mobx 6+. As described in documentation computeds "evaluate lazily, caching their output and only recomputing if one of the underlying observables has changed". So if we have a code like this:

```js
class Counter {
  value = 1;

  constructor() {
    makeAutoObservable(this)
  }

  increment() {
    this.value++;
  }
  
  get isDividedBy3() {
    return this.value % 3 === 0;
  }
}

const counter = new Counter();
autorun(() => {
  console.log('autorun', counter.isDividedBy3);
});

counter.increment();
counter.increment();
```

The output will be the following:

```
'autorun' false
'autorun' true
```

This is expected behaviour. But if I replace `autorun` with `spy` I get unexpected log statements:

```js
spy((event) => {
  if (event.type === 'update' && event.observableKind === 'computed') {
    console.log('spy', event.oldValue, '->', event.newValue);
  }
});
```

Here the console output: 
```
spy CaughtException {cause: null, constructor: Object} -> false
spy false -> false // Why is it here?
spy false -> true
```

**Intended outcome:**

For a logging library it is expected to behave the same way as actual code does - log if computed has changed and component rerendered but log nothing if the computed hasn't changed. 

**How to reproduce the issue:**

https://codesandbox.io/s/hungry-resonance-hfegb

**Versions**

Mobx 6

**Why is it happening?**

It is happening because `spyReport` is calling before the equality check in computed: https://github.com/mobxjs/mobx/blob/main/packages/mobx/src/core/computedvalue.ts#L207+L224

My suggestion is to swap them and introduce a new property for `IComputedDidChange`:

```diff
        const changed =
            wasSuspended ||
            isCaughtException(oldValue) ||
            isCaughtException(newValue) ||
            !this.equals_(oldValue, newValue)

        if (__DEV__ && isSpyEnabled()) {
            spyReport({
                observableKind: "computed",
                debugObjectName: this.name_,
                object: this.scope_,
                type: "update",
                oldValue: this.value_,
                newValue,
+              changed,
            } as IComputedDidChange)
        }
```

Ready to make a PR. Changing the `spy` behavior would be a BC break.

**Comment by urugator** [comment]
[The documentation states it should behave like `observe`](https://mobx.js.org/analyzing-reactivity.html#spy). `observe` ignores transactions in case of observables and reports on every mutation - so expecting it to behave like `autorun` is not necessarily correct in this case. 
However the `observe` for `computed` is actually implemented with `autorun` and note the comment in the sources 😅:
https://github.com/mobxjs/mobx/blob/8a1ff856043d59396f623f8ca209822b1331d85f/packages/mobx/src/core/computedvalue.ts#L270

The problem is, `spy` is global and can't attach `autorun` to every `computed` - which means that `spy` doesn't report for computeds that are not observed by something else. Also, because there is no `autorun` associated with `spy`, it doesn't see transaction boundary, therefore you see multiple reports (`false -> false`). Imo this is a correct behavior, because it matches
`observe`, which, as already noted, ignores transactions.
The thing that is imo implemented incorrectly here is `observer(computed)` - it shouldn't use `autorun` - but then it wouldn't report unless there is some other `autorun/reaction` ... which would be probably confusing to user.
So I would probably suggest to use `autorun` just to keep the `computed` alive, but invoke the actual listener in the same place as `spyReport`. That I think would result in the most consistent behavior.  @mweststrate  Thoughts?

**Comment by urugator** [comment]
Since `observe` never reports if observable value doesn't change I think we can safely move `spyReport` behind the `if (changed)` check, this is just a bug imo (wouldn't worry about BC).
Maybe the behavior will then actually match `observe(computed)` and we could call it day.

**Comment by kubk** [comment]
@urugator Thank you so much for the answer. I understand 

>  The extra false -> false is probably caused by that "something else" (autorun) pulling the value from computed for the first time.

That `false -> false` remains visible even if I remove `autorun` call: https://codesandbox.io/s/hungry-resonance-hfegb?file=/index.js

Moving `spyReport` behind the `if` would solve the problem.

**Comment by urugator** [comment]
> That false -> false remains visible even if I remove autorun call: https://codesandbox.io/s/hungry-resonance-hfegb?file=/index.js

Yes, there is still the `observer`, it reads for the first time causing `CaughtException -> false` 
I haven't noticed there is no `action` around these  `counter.increment(); counter.increment();`. Therefore 2 computations causing `false -> false` and `false -> true` ... the autorun is obviously notified only once on actual change...

> Moving spyReport behind the if would solve the issue.

I would tell you to go ahead and PR, but you may want to wait for confirmation from @mweststrate
